                    ###########################################
                    ############___ELGHAZI_SOUFIANE___#########
                    ###########################################

## Binary Classification Lab

 ###            1. Load and Prepare the Dataset: 

In [1]:
from sklearn.datasets import load_breast_cancer 
from sklearn.model_selection import train_test_split
import pandas as pd
##
from sklearn.metrics import classification_report

#####  a. Load the Wisconsin breast cancer dataset into a pandas dataframe

In [2]:
data = load_breast_cancer()
x, y = data.data, data.target

####  b.Split the dataset into training and testing sets using the train_test_split()

In [3]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=5)

df = pd.DataFrame(x_train, columns=data.feature_names)
df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0         13.170         18.66           85.98      534.6          0.11580   
1         12.340         12.27           78.94      468.5          0.09003   
2         13.980         19.62           91.12      599.5          0.10600   
3          9.504         12.44           60.34      273.9          0.10240   
4         12.880         28.92           82.50      514.3          0.08123   
..           ...           ...             ...        ...              ...   
393       13.000         21.82           87.50      519.8          0.12730   
394       13.800         15.79           90.43      584.1          0.10070   
395       17.910         21.02          124.40      994.0          0.12300   
396       15.780         22.91          105.70      782.6          0.11550   
397        9.876         17.27           62.92      295.4          0.10890   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.12310         0.12260              0.07340         0.2128   
1             0.06307         0.02958              0.02647         0.1689   
2             0.11330         0.11260              0.06463         0.1669   
3             0.06492         0.02956              0.02076         0.1815   
4             0.05824         0.06195              0.02343         0.1566   
..                ...             ...                  ...            ...   
393           0.19320         0.18590              0.09353         0.2350   
394           0.12800         0.07789              0.05069         0.1662   
395           0.25760         0.31890              0.11980         0.2113   
396           0.17520         0.21330              0.09479         0.2096   
397           0.07232         0.01756              0.01952         0.1934   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.06777  ...         15.67          27.95   
1                   0.05808  ...         13.61          19.27   
2                   0.06544  ...         17.04          30.80   
3                   0.06905  ...         10.23          15.66   
4                   0.05708  ...         13.89          35.74   
..                      ...  ...           ...            ...   
393                 0.07389  ...         15.49          30.73   
394                 0.06566  ...         16.57          20.86   
395                 0.07115  ...         20.80          27.78   
396                 0.07331  ...         20.19          30.50   
397                 0.06285  ...         10.42          23.22   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             102.80       759.4            0.1786             0.4166   
1              87.22       564.9            0.1292             0.2074   
2             113.90       869.3            0.1613             0.3568   
3              65.13       314.9            0.1324             0.1148   
4              88.84       595.7            0.1227             0.1620   
..               ...         ...               ...                ...   
393           106.20       739.3            0.1703             0.5401   
394           110.30       812.4            0.1411             0.3542   
395           149.60      1304.0            0.1873             0.5917   
396           130.30      1272.0            0.1855             0.4925   
397            67.08       331.6            0.1415             0.1247   

     worst concavity  worst concave points  worst symmetry  \
0            0.50060               0.20880          0.3900   
1            0.17910               0.10700          0.3110   
2            0.40690               0.18270          0.3179   
3            0.08867               0.06227          0.2450   
4            0.24390               0.06493          0.2372   
..               ...                   ...             ...   
393          0.53900               0.20600         

#### c. Print the number of instances and features in the training and testing sets.

In [4]:
print("  --Number of instances and features in the training set:")
print("------------------------------------------------------")
print("Training instances:", x_train.shape[0])
print("Training features:", x_train.shape[1])
print("______________________________________________________")
print("\n  --Number of instances and features in the testing set:")
print("---------------------------------------------------------")
print("Testing instances:", x_test.shape[0])
print("Testing features:", x_test.shape[1])

  --Number of instances and features in the training set:
------------------------------------------------------
Training instances: 398
Training features: 30
______________________________________________________

  --Number of instances and features in the testing set:
---------------------------------------------------------
Testing instances: 171
Testing features: 30


###            2. Train and Test Multiple Classifiers: 

#### a. Train a KNN classifier using the KNeighborsClassifier() function

In [5]:
from sklearn.neighbors import KNeighborsClassifier
model_1 = KNeighborsClassifier(n_neighbors=5)
model_1.fit(x_train, y_train)

KNeighborsClassifier()

#### b. Train a Decision Tree classifier using the DecisionTreeClassifier() function

In [6]:
from sklearn.tree import DecisionTreeClassifier
model_2 = DecisionTreeClassifier(random_state=42)
model_2.fit(x_train, y_train)

DecisionTreeClassifier(random_state=42)

#### c. Train a Logistic Regression classifier using the LogisticRegression() function 

In [7]:
from sklearn.linear_model import LogisticRegression
model_3 = LogisticRegression(random_state=0)
model_3.fit(x_train, y_train)

C:\Users\elgha\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=0)

####  d. Train a Random Forest classifier using the RandomForestClassifier() function

In [8]:
from sklearn.ensemble import RandomForestClassifier
model_4 = RandomForestClassifier(n_estimators=100, random_state=0)
model_4.fit(x_train, y_train)

RandomForestClassifier(random_state=0)

#### e. Train a Support Vector Machine (SVM) classifier using the SVC() function

In [9]:
from sklearn.svm import LinearSVC
model_5 = LinearSVC(random_state=0, tol=1e-5)
model_5.fit(x_train, y_train)

C:\Users\elgha\anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


LinearSVC(random_state=0, tol=1e-05)

#### f. Evaluate the performance of each classifier on the testing set

In [10]:
from sklearn.metrics import classification_report
target_names = ['class 1', 'class 2']

                        ###########################################
                        ############___KNeighborsClassifier___#####
                        ###########################################
y_predec_model_1 =model_1.predict(x_test)
print("                     ___KNeighborsClassifier___")            
print(classification_report(y_test, y_predec_model_1, target_names=target_names))

                     ___KNeighborsClassifier___
              precision    recall  f1-score   support

     class 1       0.96      0.89      0.92        61
     class 2       0.94      0.98      0.96       110

    accuracy                           0.95       171
   macro avg       0.95      0.93      0.94       171
weighted avg       0.95      0.95      0.95       171



In [11]:
                        ###########################################
                        ############___DecisionTreeClassifier___###
                        ###########################################
y_predec_model_2 =model_2.predict(x_test)
print("                     ___DecisionTreeClassifier___")            
print(classification_report(y_test, y_predec_model_2, target_names=target_names))

                     ___DecisionTreeClassifier___
              precision    recall  f1-score   support

     class 1       0.92      0.92      0.92        61
     class 2       0.95      0.95      0.95       110

    accuracy                           0.94       171
   macro avg       0.94      0.94      0.94       171
weighted avg       0.94      0.94      0.94       171



In [12]:
                        ###########################################
                        ############___LogisticRegression___#######
                        ###########################################
y_predec_model_3 =model_3.predict(x_test)
print("                     ___LogisticRegression___")            
print(classification_report(y_test, y_predec_model_3, target_names=target_names))

                     ___LogisticRegression___
              precision    recall  f1-score   support

     class 1       0.98      0.95      0.97        61
     class 2       0.97      0.99      0.98       110

    accuracy                           0.98       171
   macro avg       0.98      0.97      0.97       171
weighted avg       0.98      0.98      0.98       171



In [13]:
                        ###########################################
                        ############___RandomForestClassifier___###
                        ###########################################
y_predec_model_4 =model_4.predict(x_test)
print("                     ___RandomForestClassifier___")            
print(classification_report(y_test, y_predec_model_4, target_names=target_names))

                     ___RandomForestClassifier___
              precision    recall  f1-score   support

     class 1       0.97      0.95      0.96        61
     class 2       0.97      0.98      0.98       110

    accuracy                           0.97       171
   macro avg       0.97      0.97      0.97       171
weighted avg       0.97      0.97      0.97       171



In [14]:
                        ###########################################
                        ####___Support Vector Machine (SVM) ___####
                        ###########################################
y_predec_model_5 =model_5.predict(x_test)
print("               ___Support Vector Machine (SVM)___")            
print(classification_report(y_test, y_predec_model_4, target_names=target_names))

               ___Support Vector Machine (SVM)___
              precision    recall  f1-score   support

     class 1       0.97      0.95      0.96        61
     class 2       0.97      0.98      0.98       110

    accuracy                           0.97       171
   macro avg       0.97      0.97      0.97       171
weighted avg       0.97      0.97      0.97       171



###                    3. Tune Hyperparameters: 

##### a. Use GridSearchCV to find the best hyperparameters for the KNN classifier.Vary the number of neighbors from 1 to 10 and use 5-fold cross-validation. 


In [15]:
from sklearn.model_selection import GridSearchCV
param_grid = {'n_neighbors': range(1, 11)}
grid_search = GridSearchCV(model_1, param_grid, cv=5)

grid_search.fit(x_train, y_train)

print("                     ___KNeighborsClassifier___") 
print("______________________________________________________") 
print("Best hyperparameters: ", grid_search.best_params_)
print("Best accuracy score: ", grid_search.best_score_)

                     ___KNeighborsClassifier___
______________________________________________________
Best hyperparameters:  {'n_neighbors': 10}
Best accuracy score:  0.9220569620253165


#### b. Use GridSearchCV to find the best hyperparameters for the Decision Tree classifier. Vary the maximum depth from 1 to 10 and use 5-fold cross_validation.


In [16]:
param_grid = {'max_depth': range(1, 11)}
grid_search = GridSearchCV(model_2, param_grid, cv=5)

grid_search.fit(x_train, y_train)
print("                     ___DecisionTreeClassifier___") 
print("______________________________________________________")
print("Best hyperparameters: ", grid_search.best_params_)
print("Best accuracy score: ", grid_search.best_score_)

                     ___DecisionTreeClassifier___
______________________________________________________
Best hyperparameters:  {'max_depth': 3}
Best accuracy score:  0.9295886075949367


#### c. Use GridSearchCV to find the best hyperparameters for the Logistic Regression classifier. Vary the regularization parameter C from 0.01 to 10 and use 5-fold cross-validation.

In [17]:
param_grid = {'C': [0.01, 0.1, 1, 10]}
grid_search = GridSearchCV(model_3, param_grid, cv=5)

grid_search.fit(x_train, y_train)
print("                     ___LogisticRegression___")
print("______________________________________________________")
print("Best hyperparameters: ", grid_search.best_params_)
print("Best accuracy score: ", grid_search.best_score_)

C:\Users\elgha\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\elgha\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_

                     ___LogisticRegression___
______________________________________________________
Best hyperparameters:  {'C': 1}
Best accuracy score:  0.9421518987341774


C:\Users\elgha\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\elgha\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_

#### d. Use GridSearchCV to find the best hyperparameters for the Random Forest classifier. Vary the number of trees from 50 to 200 and use 5-fold cross-validation. 

In [18]:
param_grid = {'n_estimators': range(50, 201)}
grid_search = GridSearchCV(model_4, param_grid, cv=5)

grid_search.fit(x_train, y_train)
print("                     ___RandomForestClassifier___")  
print("______________________________________________________")
print("Best parameters: ", grid_search.best_params_)
print("Best accuracy score: ", grid_search.best_score_)

                     ___RandomForestClassifier___
______________________________________________________
Best parameters:  {'n_estimators': 55}
Best accuracy score:  0.9623101265822787


#### e. Use GridSearchCV to find the best hyperparameters for the SVM classifier. Vary the regularization parameter C from 0.01 to 10 and use 5-fold cross-validation.

In [19]:
param_grid = {'C': [0.01, 0.1, 1, 10]}
grid_search = GridSearchCV(model_5, param_grid, cv=5)

grid_search.fit(x_train, y_train)
print("               ___Support Vector Machine (SVM)___") 
print("______________________________________________________")
print("Best parameters: ", grid_search.best_params_)
print("Best accuracy score: ", grid_search.best_score_)

C:\Users\elgha\anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\elgha\anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\elgha\anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\elgha\anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\elgha\anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\elgha\anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number 

               ___Support Vector Machine (SVM)___
______________________________________________________
Best parameters:  {'C': 10}
Best accuracy score:  0.901867088607595


C:\Users\elgha\anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


#### f. Print the best hyperparameters and the corresponding performance metrics for each classifier.

### 4. Compare and Analyze Results:

#### a. Compare the performance of each classifier before and after hyperparameter tuning.

    En se basent sur les résultats affichés, on trouve que le max de l’accuracy est 98% par le classificateur LogisticRegression, suivi par RandomForestClassifier et SVM de précision 97%, puis KNeighborsClassifier 95% et enfin DecisionTreeClassifier 94%.

#### b. Analyze the best hyperparameters for each classifier and their impact on the performance metrics. 

    Après Tuning, on remarque que l’hyperparamètres ont affecté la précision score telle que RandomForestClassifier qui a la valeur maximale 96,23%, suivi par LogisticRegression 94,21%, puis KNeighborsClassifier et DecisionTreeClassifier par 92% et enfin SVM 90,18%.
    Alors on remarque que l’ordre a changé ainsi que la précision à diminuer !


#### c. Discuss the strengths and weaknesses of each classifier and their suitability for this particular dataset.

###### 	KNeighborsClassifier:
Le KNN utilise la classification de voisinage comme valeur de prédication de la nouvelle requête. Il a des avantages - l'architecture non paramétrique, simple et puissante, ne nécessite aucun temps d'apprentissage, mais il a aussi un inconvénient - gourmand en mémoire, la classification et l'estimation sont lentes.
###### 	DecisionTreeClassifier :
Avantages des arbres de décision : Interprétabilité. Moins de préparation des données. Non paramétrique. Polyvalence. Non-linéarité.
Inconvénients de l'arbre de décision : Surajustement. Réduction des fonctionnalités et rééchantillonnage des données. Optimisation.
###### 	LogisticRegression
Avantages : Algorithme d'apprentissage automatique le plus simple et Moins sujet au surajustement et Plus précise
Inconvénients : Surajustement, Problème avec les relations complexes 
###### 	RandomForestClassifier
Avantages : Parmi toutes les méthodes de classification disponibles, les forêts aléatoires offrent la plus grande précision. La technique de la forêt aléatoire peut également gérer des données volumineuses avec de nombreuses variables se comptant en milliers. Il peut automatiquement équilibrer les ensembles de données lorsqu'une classe est moins fréquente que d'autres classes dans les données
Inconvénients : La principale limitation de la forêt aléatoire est qu'un grand nombre d'arbres peut rendre l'algorithme trop lent et inefficace pour les prédictions en temps réel. En général, ces algorithmes sont rapides à former, mais assez lents à créer des prédictions une fois qu'ils sont formés.
###### 	SVC
Avantages : SVM fonctionne relativement bien lorsqu'il existe une marge de séparation claire entre les classes. SVM est plus efficace dans les espaces de grande dimension et est relativement économe en mémoire. SVM est efficace dans les cas où les dimensions sont supérieures au nombre d'échantillons.
Inconvénients : Il ne fonctionne pas bien lorsque nous avons un grand ensemble de données car le temps de formation requis est plus élevé.
Il ne fonctionne pas non plus très bien lorsque l'ensemble de données a plus de bruit, c'est-à-dire que les classes cibles se chevauchent.
